tests/integration/test_upgrade.py

```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import asyncio
+import logging
+import time
+from pathlib import Path
+
+import pytest
+import yaml
+from pytest_operator.plugin import OpsTest
+
+from literals import DEPENDENCIES
+
+from .helpers import correct_version_running, get_relation_data, ping_servers
+
+logger = logging.getLogger(__name__)
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+APP_NAME = METADATA["name"]
+
+# FIXME: update this to 'stable' when `pre-upgrade-check` is released to 'stable'
+CHANNEL = "edge"
+
+
+@pytest.mark.abort_on_fail
+async def test_in_place_upgrade(ops_test: OpsTest):
+    build_charm = asyncio.ensure_future(ops_test.build_charm("."))
+
+    await ops_test.model.deploy(APP_NAME, application_name=APP_NAME, num_units=3, channel=CHANNEL)
```

Comment:
Yes, I definitely agree with Raul (or actually I agree with both). 

I believe we should pin a version (or versions, making this parametric) of our choice to ensure we can upgrade from a given revision (possibly the latest stable ones). Also the tests against the channels "edge" can be sensible, but I would have this on top 

My idea would be to have this set to the stable revisions that we say we support upgrade from. It would probably be also good to write some data and then make sure we can read the data during/after upgrade.

Reply:
I'm not sure I agree.

We are using `X/stable` channels for minor versions, which is what we should test against. So for ZooKeeper, that would be `3/stable` right now, which is what we'll pin in L23/24.

We can parameterize this when we have more than `3/stable`, but isn't necessary work now.
We also aren't published to `stable` yet, which is why we can't use that now.

Pinning against revisions seems like a huge headache. We instead choose to prioritise upgrading from any `stable` to any `stable`. Once we publish to `stable`, then it makes sense to periodically update the pinned revisions list every time we modify `DEPENDENCIES`, but for the sake of this PR, we don't have that yet.

Reply:
That's fine. In general, I believe we should allow to upgrade from few stable revisions (we could make something like the 3 latest stable revisions), unless this is not possible (e.g. because of some breaking changes), and clarify this in the rollout guide. But it is probably something to also keep consistent across DP. 

Fine to make it out-of-scope for now